cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringServerHandler.java

```diff
@@ -0,0 +1,265 @@
+    TopicId topic = new TopicId(NamespaceId.SYSTEM.getNamespace(), topicPrefix + peer);
+    String lastMessageId = messageId;
+    try (CloseableIterator<Message> iterator =
+           fetcher.fetch(topic.getNamespace(), topic.getTopic(), 1, messageId)) {
+      while (iterator.hasNext()) {
+        Message message = iterator.next();
+        TetheringControlMessage controlMessage = GSON.fromJson(message.getPayloadAsString(StandardCharsets.UTF_8),
+                                                               TetheringControlMessage.class);
+        commands.add(controlMessage);
+        lastMessageId = message.getId();
+      }
+    } catch (TopicNotFoundException e) {
+      LOG.warn("Received control connection from peer {} that's not tethered", peer);
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(String.format("Invalid message id %s", messageId));
+    }
+
+    if (commands.isEmpty()) {
+      commands.add(new TetheringControlMessage(TetheringControlMessage.Type.KEEPALIVE));
+    }
+    TetheringControlResponse tetheringControlResponse = new TetheringControlResponse(lastMessageId, commands);
+    final InputStream in = new ByteArrayInputStream(GSON.toJson(tetheringControlResponse).getBytes());
+    try {
+      responder.sendContent(HttpResponseStatus.OK, new BodyProducer() {
+
+        @Override
+        public ByteBuf nextChunk() throws Exception {
+          ByteBuf buffer = Unpooled.buffer(MAX_CHUNK_SIZE);
+          buffer.writeBytes(in, MAX_CHUNK_SIZE);
+          return buffer;
+        }
+
+        @Override
+        public void finished() {
+          Closeables.closeQuietly(in);
+        }
+
+        @Override
+        public void handleError(@Nullable Throwable cause) {
+          Closeables.closeQuietly(in);
+        }
+      }, new DefaultHttpHeaders().add(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_OCTET_STREAM));
+    } catch (Exception e) {
+      Closeables.closeQuietly(in);
+      throw e;
+    }
+  }
+
+  /**
+   * Creates a tethering with a client.
+   */
+  @POST
+  @Path("/tethering/connect")
+  public void createTethering(FullHttpRequest request, HttpResponder responder)
+    throws NotImplementedException, IOException {
+    checkTetheringServerEnabled();
+
+    String content = request.content().toString(StandardCharsets.UTF_8);
+    TetheringConnectionRequest tetherRequest = GSON.fromJson(content, TetheringConnectionRequest.class);
+    TopicId topicId = new TopicId(NamespaceId.SYSTEM.getNamespace(),
+                                  topicPrefix + tetherRequest.getPeer());
+    try {
+      messagingService.createTopic(new TopicMetadata(topicId, Collections.emptyMap()));
+    } catch (TopicAlreadyExistsException e) {
+      LOG.warn("Topic {} already exists", topicId);
+    } catch (IOException e) {
+      LOG.error("Failed to create topic {}", topicId, e);
+      throw e;
+    }
+
+    try {
+      store.getPeer(tetherRequest.getPeer());
```

Comment:
get and add should be in the same transaction.

Reply:
fixed